Tribler/Policies/BoostingSource.py

```diff
@@ -0,0 +1,639 @@
+                else:
+                    for candidate in torrent.query_candidates:
+                        self.session.download_torrentfile_from_peer(candidate, torrent.infohash, None, 0)
+
+                torrent = NotCollectedTorrent(torrent, files, trackers)
+
+            else:
+                tdef = TorrentDef.load_from_memory(self.session.get_collected_torrent(torrent.infohash))
+
+                if torrent.torrent_id <= 0:
+                    del torrent.torrent_id
+
+                torrent = CollectedTorrent(torrent, tdef)
+
+        # replacement # self.library_manager.addDownloadState(torrent)
+        for dl_state in self.dslist:
+            torrent.addDs(dl_state)
+
+        # return
+        if callback is not None:
+            callback(torrent)
+        else:
+            return torrent
+
+    @staticmethod
+    def get_instance(*args, **kw):
+        """
+        get single instance of TorrentManagerCM
+        """
+        if TorrentManagerCM.__single is None:
+            TorrentManagerCM(*args, **kw)
+        return TorrentManagerCM.__single
+
+    @staticmethod
+    def del_instance():
+        """
+        resetting, then deleting single instance
+        """
+        TorrentManagerCM.__single = None
+
+
+class BoostingSource(object):
+    """
+    Base class for boosting source. For now, it can be RSS, directory, and channel
+    """
+
+    def __init__(self, session, source, boost_settings, callback):
+        self.session = session
+        self.channelcast_db = session.lm.channelcast_db
+
+        self.torrents = {}
+        self.source = source
+        self.interval = boost_settings.source_interval
+        self.max_torrents = boost_settings.max_torrents_per_source
+        self.callback = callback
+        self.archive = False
+
+        self.enabled = True
+
+        self.av_uprate = 0
+        self.av_dwnrate = 0
+        self.storage_used = 0
+        self.ready = False
+
+        self.min_connection = boost_settings.min_connection_start
+        self.min_channels = boost_settings.min_channels_start
+
+        self.torrent_mgr = TorrentManagerCM.get_instance(session)
+
+        # local import for handling circular import
+        from Tribler.Policies.BoostingManager import BoostingManager
+        self.boosting_manager = BoostingManager.get_instance()
+        self._logger = logging.getLogger(BoostingManager.__name__)
+
+    def kill_tasks(self):
+        """
+        kill tasks on this source
+
+        *not used right now*
+        """
+        # cancel loading task
+
+        self.ready = False
+        self.torrent_mgr.del_instance()
+
+        self.session.lm.threadpool.cancel_pending_task(str(self.source) + "_load")
+
+        self.session.lm.threadpool.cancel_pending_task(self.source)
+
+    def _load_if_ready(self, source):
+        """
+        load source if and only if the overall system is ready.
+
+        it depends on #connection and #channel
```

Comment:
yes, in `BoostingSettings` we can define this. The default value is 5 connection and 100 channels.